KEP-6032: Add initial proposal (localhost NodePort userspace proxy) (#6042)
kubernetes/enhancements: keps/sig-network/6032-localhost-nodeport-userspace-proxy/README.md

  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
- [ ] (R) Graduation criteria is in place
  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) within one minor version of promotion to GA
- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

> This KEP proposes implementing a userspace proxy to enable localhost NodePort services across all proxy backends. The iptables proxy already works with localhost NodePorts; however, it requires enabling a dangerous permission, `route_localnet`, on each node to allow the kernel to rewrite traffic to localhost. By instead rewriting traffic directly at the userspace level, no extra permissions are set, and localhost NodePorts will work on all proxy backends and IPv6.

Reviewer: Question for sig leads: do we want to add this feature to IPVS, given that it's deprecated?

Author: I was also curious on this. I'm fine with the decision either way

Reviewer 2: I guess there are 3 possibilities:
1. Implement it in nftables, iptables, and ipvs, because it's easy and we can.
2. Implement it in nftables and iptables, but not ipvs, because ipvs is deprecated and this will provide another reason for people to migrate from ipvs to nftables.
3. Implement it only in nftables, not ipvs for the above reasons, and not iptables because iptables users can already do it with `route_localnet`, and having a more secure way to do it in nftables gives another reason for iptables users to migrate to nftables. (And then we don't have to figure out any complicated rules about when we enable `route_localnet` vs when we use the userspace proxy for iptables.)

Though option 3 leaves things broken for iptables IPv6 users... Plus, we definitely want to deprecate setting `route_localnet`, and that's easier to do if we support userspace localhost nodeports in iptables mode. So no option 3.

I can see the arguments for or against supporting ipvs. I guess one easy argument is "he already wrote the code". Also, ipvs has gotten some other new features (like Traffic Distribution) just as a side effect of there being shared kube-proxy code, so this isn't totally weird.

Also, it saves us from having to add rules to the ipvs CI job to skip this e2e test. 🙂 

So option 1?

### Goals

- Enable localhost NodePort across all proxies and IPv6 without `route_localnet`
- Stop enabling `route_localnet`

### Non-Goals

- Implement UDP or other non-TCP protocol in the userspace proxy

## Proposal

Add a userspace TCP-only proxy to "kube-proxy" that listens on the loopback address (`127.0.0.1` / `::1`) for each NodePort and forwards connections to a service's endpoints. Each backend proxy (`iptables`, `nftables`, `ipvs`) instantiates this proxy when `--nodeport-addresses` contains a loopback CIDR, and reconciles its set of listeners on every sync.

> `iptables` and `ipvs` default to `0.0.0.0/0` or `::/0`, allowing all IPs, including localhost, when `--nodeport-addresses` is unset. `nftables` currently defaults to `primary`, accepting connections only to the node IP. `nftables` will begin defaulting to `primary,<loopback-address>` to include localhost. The `--nodeport-addresses` flag will change to allow using the `primary` keyword alongside IP addresses. 

Author: Changing the default of `nftables` might be controversial. LMK thoughts @danwinship

Reviewer 2: I think we should add `localhost` as a special keyword too, so you can just say `--nodeport-addresses primary,localhost` rather than having to write out `127.0.0.1/32` (or `::1/128`, or `127.0.0.1/32,::1/128`)

But I think we need to avoid automatically enabling localhost nodeports in any cluster where they were not previously enabled. In particular, people in clusters that don't natively support localhost nodeports may currently be deploying various hacks to work around that (eg their own iptables rules, their own userspace proxy, etc). If we suddenly start intercepting packets in cases where we weren't doing that before, that may break things for some users.

So I think that:
- for iptables IPv4, we should change _how_ we implement localhost nodeports in some clusters (see below), but we shouldn't change _whether_ we implement localhost nodeports in any clusters.
- for iptables IPv6, ipvs, and nftables, we should continue to _not_ support localhost nodeports in clusters that have the default `--nodeport-addresses` value.

OTOH, right now, if you do `--iptables-localhost-nodeports=false --nodeport-addresses=xxx,yyy,127.0.0.1/32` in IPv4 iptables, or `--nodeport-addresses=xxx,yyy,127.0.0.1/32` in IPv4 nftables, or `--nodeport-addresses=xxx,yyy,::1/128` in IPv6 iptables or nftables, then kube-proxy will spew a completely excessive number of warnings to the logs. So we can safely assume that nobody is doing that, so I think it's safe to change the semantics of that.

So: for all cases except iptables IPv4, if you want localhost nodeports, you have to explicitly request it via `--nodeport-addresses`.

Author: I believe that if we use the default value of `--nodeport-addresses` to determine when to use the proxy then localhost Nodeport work by default on IPV6 IP tables and IPVS. Since the default IP address for these proxies is the zero CIDR it would include loopback. We could exclude the zero CIDR from loopback, but this feels inaccurate.

Author: I updated the proposal so that `--nodeport-addresses` has a new keyword `localhost`. nftables default behavior will not change. However, the default behavior of iptables IPv6 and IPVS would change given that their nodeport addresses already include loopback. LMK if this is acceptable or if we must introduce another flag or some other construct to avoid this. 

Could be worth re-evaluting if the IPVS changes are worth doing if we are worried about breaking changes on that side.

Reviewer 2: The existing proxy implementations already check whether you're allowing nodeports on _all_ IPs (`proxier.nodePortAddresses.MatchAll()`), and treat that differently from allowing nodeports on _specific_ IPs. (The "completely excessive number of warnings" mentioned above only happen if you _specifically_ allow nodeports on localhost, not if you allow nodeports on all IPs.) So we can still preserve the default behavior, by only using the new code if you aren't listening on all IPs.

Author: Gotcha, updated the proposal so that the userspace proxy will not be created unless `--nodeport-addresses` is not match all and loopback is set (except for iptables ipv4 once we switch over)

  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
- [ ] (R) Graduation criteria is in place
  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) within one minor version of promotion to GA
- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

> This KEP proposes implementing a userspace proxy to enable localhost NodePort services across all proxy backends. The iptables proxy already works with localhost NodePorts; however, it requires enabling a dangerous permission, `route_localnet`, on each node to allow the kernel to rewrite traffic to localhost. By instead rewriting traffic directly at the userspace level, no extra permissions are set, and localhost NodePorts will work on all proxy backends and IPv6.

Reviewer 2: please break lines to make it easier to review to individual pieces

Explain IPv6. (You mention your implementation will make this work on IPv6 but you didn't mention that IPv6 doesn't work before that.)

The discussion here is Linux-specific and you should make that clear. (You say "all proxy backends" but I'm pretty sure you're not planning to fix it on `winkernel` too. I have no idea whether localhost nodeports current work on Windows or not, but if they don't, it's not because of `route_localnet`.)

## Motivation

> `route_localnet` is a dangerous kernel permission set on every node node using IPv4 and iptables by default. Implementing a userspace proxy will enable a secure default without breaking existing use cases. See [#132955](https://github.com/kubernetes/kubernetes/issues/132955).

Reviewer 2: ```suggestion
`route_localnet` is a dangerous kernel permission that kube-proxy in iptables mode sets on every node node using IPv4 by default. Implementing a userspace proxy will enable a secure default without breaking existing use cases. See [#132955](https://github.com/kubernetes/kubernetes/issues/132955).
```

> In-cluster registries in disconnected environments are a popular production use case for localhost NodePorts. The tool [Zarf](https://github.com/zarf-dev/zarf) has a built-in process to start up a registry served over a localhost NodePort on an airgapped cluster. The process is distro and container runtime interface (CRI) agnostic. This is possible because popular CRIs such as containerd and cri-o allow insecure localhost connections to registries. Without this, users would need to edit their host node configuration to either include TLS certificates or allow insecure connections to node IPs. Enabling localhost NodePorts makes it easier for these users to move from iptables to nftables.

Reviewer 2: need to clarify that the problem is that zarf only works with the iptables mode of kube-proxy. I tried "The process is distro and container runtime interface (CRI) agnostic, but only works with the iptables mode of kube-proxy", but then the next sentence doesn't flow.

This paragraph would work better as a User Story anyway

### Goals

> - Enable localhost NodePort across all proxies and IPv6 without `route_localnet`

Reviewer 2: ```suggestion
- Enable localhost NodePort across all Linux kube-proxy modes, for both IPv4 and IPv6, without `route_localnet`
```

- Stop enabling `route_localnet`

### Non-Goals

- Implement UDP or other non-TCP protocol in the userspace proxy

## Proposal

Add a userspace TCP-only proxy to "kube-proxy" that listens on the loopback address (`127.0.0.1` / `::1`) for each NodePort and forwards connections to a service's endpoints. Each backend proxy (`iptables`, `nftables`, `ipvs`) instantiates this proxy when `--nodeport-addresses` contains a loopback CIDR, and reconciles its set of listeners on every sync.

`iptables` and `ipvs` default to `0.0.0.0/0` or `::/0`, allowing all IPs, including localhost, when `--nodeport-addresses` is unset. `nftables` currently defaults to `primary`, accepting connections only to the node IP. `nftables` will begin defaulting to `primary,<loopback-address>` to include localhost. The `--nodeport-addresses` flag will change to allow using the `primary` keyword alongside IP addresses. 

> This unblocks localhost NodePorts for `nftables` (IPv4 and IPv6), `ipvs` (IPv4 and IPv6), and `iptables` IPv6. `iptables` IPv4 will continue to use `route_localnet` until this feature graduates to GA, then transition to the userspace proxy. The field `--iptables-localhost-nodeports` will be deprecated in favor of explicitly setting `--nodeport-addresses`.

Reviewer 2: So, the goals here are:
1. Stop setting `route_localnet` by default, because security.
2. Don't break existing configurations, except as required by point 1, and in that case, phase in the breakage properly.

Theoretically, switching existing `route_localnet` users over to the userspace proxy may not completely avoid breaking existing configurations. (e.g., they may depend on UDP localhost nodeports). I'm not sure how theoretical that is and how much we care.

If we care, and we want to continue supporting the `route_localnet` hack forever, for full backward compatibility, the approach should be:

1. Add a new config field/CLI arg `--iptables-set-route-localnet` which is just an alias for `--iptables-localhost-nodeports`. Add a new feature gate `KubeProxySetRouteLocalnet` which is initially `Default: true, PreRelease: featuregate.GA`, and have the default value of `--iptables-set-route-localnet` be the value of the feature gate. If the user explicitly sets `--iptables-localhost-nodeports=false`, log a warning that that name is deprecated in favor of the new name.
2. Implement userspace localhost nodeports.
    - When `--iptables-set-route-localnet` is `true` (initially the default), ignore the new userspace localhost nodeport code and continue doing everything exactly like we used to, but log a warning that `--iptables-set-route-localnet` is insecure and deprecated and will default to `false` in the future, and explaining how to enable non-`route_localnet`-based nodeports. (Yes, it's a long warning message!)
    - When `--iptables-set-route-localnet` is `false`, decide whether to use the userspace localhost nodeport code or not depending on `--nodeport-addresses`, as above.
3. Eventually change the `KubeProxySetRouteLocalnet` feature gate to `Default: false, PreRelease: featuregate.Deprecated`, and then eventually also set `LockToDefault: true`
    - Probably at the same time remove `--iptables-localhost-nodeports` and require people to use `--iptables-set-route-localnet` instead.
    - Though not entirely related, this is probably also a good time to finally make kubelet stop creating a duplicate copy of the `route_localnet`-mitigation firewall rule, as [discussed in KEP-3178](https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/3178-iptables-cleanup/README.md#martian-packet-blocking).

If we _don't_ want to continue supporting `route_localnet` forever, then we can probably include deprecating `--iptables-localhost-nodeports` as part of the `LocalhostNodePortUserspaceProxy` feature gate.

Author: My opinion is that it makes more sense to deprecate the route_localnet path and when/if people complain that the userspace proxy is missing functionality add to the proxy. I'll leave the final decision to you.

### User Stories

#### Story 1

A cluster operator migrating from `iptables` to `nftables` discovers their workload depends on localhost NodePorts. Today the migration breaks that workload; with userspace localhost NodePorts it keeps working.

### Risks and Mitigations

Some operators may not want localhost NodePort to be reachable in their cluster. These operators can set `--nodeport-addresses=primary`.

## Design Details

Design details are covered in the proposal section.

### Test Plan

[X] I/we understand the owners of the involved components may require updates to
existing tests to make this code solid enough prior to committing the changes necessary
to implement this enhancement.

##### Prerequisite testing updates

N/A

##### Unit tests

> All code should be covered.

Reviewer 2: you need to include the relevant code coverage numbers, as shown in the original template text here

### Risks and Mitigations

The proxy is opt-in for nftables and does not change any defaults, so risks are low.

## Design Details

### Service Feature Support

The userspace localhost NodePort proxy is a Layer 4 TCP forwarder. The following covers Service spec fields and how the proxy handles them.

**Implemented:**

- `protocol: TCP` — the proxy creates a listener for every TCP NodePort included in the sync.
- `sessionAffinity: ClientIP` — the proxy pins traffic to a single endpoint for `sessionAffinityConfig.clientIP.timeoutSeconds`. Because all connections originate from `127.0.0.1`/`::1`, ClientIP affinity effectively pins *all* traffic through the listener to one endpoint—not per external client—until the timeout elapses.
- `externalTrafficPolicy: Local` — the proxy restricts its endpoint pool to node-local endpoints. If no local endpoints exist the listener is removed and connections are refused.
- `trafficDistribution` — implemented, endpoints are already categorized before the userspace proxy is created.

**Not implemented:**

> - `protocol: UDP` / `protocol: SCTP` — the proxy creates no listener and logs an error. May be reconsidered if there is user demand.

Reviewer 3: one thought will be to open the socket for UDP and SCTP and immediately  close it so the kernel send an error and logs a metric ... this can provide platforms telemetry to understand if there is user demand because they can scrape those metrics

Author: I'd be open to that, another option might be to add a metric for it. Let me know if I should add it.

Reviewer 2: You wouldn't even need to create a socket; you can just do it within the nftables ruleset. Create a named counter, and then have a rule that matches UDP nodeport connections, updates the counter, and rejects the packet, and then periodically update the metric from the counter.

Author: If I'm understanding this correctly we would have still log an error when a localhost nodeport with sctp or UDP is created. Then in addition we would have a separate metric, maybe `kubeproxy_nftables_localhost_nodeport_rejected_packets_total{protocol}` so operators can have visibility into if their packets to UDP or SCTP are being rejected. Is that right?

Reviewer 2: We can either log or not log when we create the rules for the service, and independently, we can either reject or not reject when a packet comes in.

The argument against logging is that at the point when we're creating the rules, we don't _know_ that the user expects localhost nodeports to work with _that_ service, so we might be basically warning them "Something you weren't planning to do isn't going to work".

So I vote for "don't log, do reject".

- [X] Metrics
  - Metric name: `kubeproxy_nftables_localhost_nodeport_listeners`
  - Components exposing the metric: `kube-proxy`

###### Are there any missing metrics that would be useful to have to improve observability of this feature?

We could add a metric similar to the existing `kubeproxy_iptables_localhost_nodeports_accepted_packets_total` to allow operators to view how often the userspace localhost proxy is being hit.

### Dependencies

###### Does this feature depend on any specific services running in the cluster?

No.

### Scalability

###### Will enabling / using this feature result in any new API calls?

No.

###### Will enabling / using this feature result in introducing new API types?

No.

###### Will enabling / using this feature result in any new calls to the cloud provider?

No.

###### Will enabling / using this feature result in increasing size or count of the existing API objects?

No.

###### Will enabling / using this feature result in increasing time taken by any operations covered by existing SLIs/SLOs?

No.

###### Will enabling / using this feature result in non-negligible increase of resource usage (CPU, RAM, disk, IO, ...) in any components?

> No.

Reviewer 3: Yes, unless the proxy does splicing the proxy will have to hold the data transferred in memory , I think golang uses by default some buffering ... but it will interesting to know in the kube-proxy memory impact specially since this new AI images has sizes of Gbs ... we need to avoid to oom kill the proxy

Author: It looks like according to this go documentation https://go.dev/doc/go1.11#netpkgnet the linux splice call is automatically used by `io.copy` in TCP connections. Many conncurrent connections might be a worry though I'm not sure that's worth noting

Reviewer 3: just to clarify, the goal of this section is not to push back on features, is just to document the changes ... if we don't buffer then there is no ram usage, but the context switching of packets will impact cpu ... we should try to describe well the impact

Author: Gotcha, added a note about the CPU overhead

###### Can enabling / using this feature result in resource exhaustion of some node resources (PIDs, sockets, inodes, etc.)?

> A client could create many NodePort services to exhaust the node's port range, but this is already possible with NodePort services and is bounded by the operator's NodePort allocation.

Reviewer 3: All userspace proxy consume sockets, the listening one and the proxied one, that limit of socket is defined by the number of concurrent connections, so the final formula for the upper bound will be:

(number of node ports) x (number of connections per nodeport)

Author: Added a mention for sockets, though I put the equation as `2 × (total concurrent proxied connections)`. I don't think the number of nodeports should be a figure since clients are likely only connecting to one nodeport as a time. Let me know if I'm missing anything

Reviewer 3: as in https://github.com/kubernetes/enhancements/pull/6042#discussion_r3306786396, these sections are for describing the impact, not for pushing back on features ...

Author: Got it, I have a mention of the sockets, let me know if there is anything else you'd like to see here

<!--

> **Note:** When your KEP is complete, all of these comment blocks should be removed.

Reviewer 2: You can remove this comment block, and the one between the title and the toc

- Change any backend's default `--nodeport-addresses`
- Deprecate or stop setting `route_localnet`
- Implement UDP or other non-TCP protocols in the userspace proxy

## Proposal

The nftables kube-proxy backend gains a userspace TCP proxy that listens on `127.0.0.1` / `::1` for each
NodePort and forwards to the service's endpoints.

`--nodeport-addresses` is extended in two ways:

- New keywords `localhost` (loopback CIDRs for the node's IP families) and `all` (zero CIDRs).
- `primary` may now appear alongside other entries instead of being required as the sole value.

The proxy activates whenever `--nodeport-addresses` explicitly includes a loopback address. For instance, it would be activated with `primary,localhost` or `0.0.0.0/0,127.0.0.1/32`, while `all` would not activate it. nftables' default (`primary`) is unchanged.

iptables and ipvs are unchanged. iptables IPv4 continues to use `route_localnet`.

### User Stories

#### Story 1

A cluster operator migrating from iptables to nftables discovers their workload depends on localhost
NodePorts. With this feature, they set `--nodeport-addresses=primary,localhost` on nftables and the

> workload keeps working.

Reviewer 2: I think this is pretty much covered by the second user story now, so you can just remove this one

- `externalTrafficPolicy: Local` — the proxy restricts its endpoint pool to node-local endpoints. If no local endpoints exist the listener is removed and connections are refused.
- `trafficDistribution` — implemented, endpoints are already categorized before the userspace proxy is created.

**Not implemented:**

- `protocol: UDP` / `protocol: SCTP` — the proxy creates no listener and logs an error. May be reconsidered if there is user demand.

**Not applicable:**

- `internalTrafficPolicy` — governs ClusterIP access; NodePort traffic is governed by `externalTrafficPolicy`.
- Any loadBalancer-specific fields (e.g. `loadBalancerSourceRanges`, `allocateLoadBalancerNodePorts`).

### Test Plan

[X] I/we understand the owners of the involved components may require updates to
existing tests to make this code solid enough prior to committing the changes necessary
to implement this enhancement.

##### Prerequisite testing updates

N/A

##### Unit tests

> Here are the current coverage percentages on master to the relevant packages for this feature:

Reviewer 2: Are you planning to add any new unit tests?

Author: For the new package, and for changes to existing packages. Added a comment to note this

- `k8s.io/kubernetes/pkg/proxy/nftables`: `2026-05-12` - `77.5`
- `k8s.io/kubernetes/pkg/proxy/util`: `2026-05-12` - `82.7`

##### Integration tests

No integration tests to be added or updated. E2E tests are sufficient since this feature only affects kube-proxy.

##### e2e tests

> E2E tests are added under the `KubeProxyNFTablesLocalhostNodePorts` feature tag to verify that:

Reviewer 2: You meant "feature gate" again, but actually, in this case, we need a feature flag, not just a feature gate, because simply enabling the feature gate isn't enough to make the feature work; you also have to have configured the cluster in the right way.

So instead of feature-gating the tests, I'd say you should add a feature flag (to `test/e2e/feature/feature.go`) for `KubeProxyLocalhostNodePorts`, and have the new e2e test require that feature, and then we can immediately start running that test in some of the IPv4 iptables jobs, and later once the feature goes to beta, we can update the config for some of the nftables jobs and then run it there as well.

Author: Yeah I'd been using feature flag and feature gate interchangeably.

Would it be correct to say a feature gate is anything declared by `--features=` while a feature flag is a test construct to decide which tests to run ?

Seems a bit strange to me to skip the e2e tests for nftables since that means the userspace proxy won't have any e2e tests until beta, but if that's the usual pattern in kube I'm fine with that.

Reviewer 2: Feature gates control whether an alpha/beta feature is _available_ to be used, but don't automatically imply that it's _actually_ being used. (True, most people enabling the feature gate during alpha are probably doing so because they want to use the feature, but once you get to beta, the feature gate will be enabled by default, and that doesn't imply the user has actually updated `nodeport-addresses` to take advantage of the new feature. Or even that they're using `nftables` mode for that matter.)

Feature tags are specific to the test framework, and are used to mark the test as requiring that some optional/platform-specific functionality has _actually_ been enabled in this cluster.

> Seems a bit strange to me to skip the e2e tests for nftables since that means the userspace proxy won't have any e2e tests until beta, but if that's the usual pattern in kube I'm fine with that.

Yeah... there's a job that runs with all alpha feature gates enabled, but that only tests features that don't require any extra configuration / feature flags. And we don't want to enable an alpha feature on the regular `nftables` jobs because that might destabilize them. In this case, we assume that _you_ are going to be trying out the code while it's alpha, and will notice and fix any obvious major bugs, and then when we get to beta we can start testing it more extensively in CI.

Author: Got it, thanks for the explanation. Updated that section

- curling `localhost:<NodePort>` from a node's netns can reach both the local node and a remote node.
- sessionAffinity=ClientIP is honored when the userspace proxy is used.

### Graduation Criteria

#### Alpha

> - Feature implemented behind a feature flag

Reviewer 2: nitpick, but

```suggestion
- Feature implemented behind a feature gate
```

- E2E tests implemented

#### Beta

- All functionality completed

> - All security enforcement completed

Reviewer 2: What's this? You didn't mention "security enforcement" anywhere else.

Author: Sorry, copy pasted the beta criteria from the comments block. Removed

- All monitoring requirements completed
- All testing requirements completed
- All known pre-release issues and gaps resolved

#### Stable

- Any issues and gaps identified as feedback during beta are resolved

### Upgrade / Downgrade Strategy

N/A

### Version Skew Strategy

N/A

## Production Readiness Review Questionnaire

> <!--

Reviewer 2: (more comments you can delete)

###### Does enabling the feature change any default behavior?

No. The userspace proxy only activates when an operator includes `localhost` (or a literal loopback CIDR)
in `--nodeport-addresses` on nftables. iptables and ipvs are not modified by this KEP.

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

Yes. Disabling the feature gate and restarting kube-proxy stops the userspace listeners; existing in-flight connections through the proxy are closed.
Traffic that previously reached the cluster via `localhost:<NodePort>` will fail until clients fall back to a node IP, but no other workload is affected.

###### What happens if we reenable the feature if it was previously rolled back?

The proxy starts up on the next kube-proxy sync, recreates listeners for any matching NodePort services, and traffic to `localhost:<NodePort>` works again.

###### Are there any tests for feature enablement/disablement?

No since the userspace proxy is standalone and does not alter existing data.

### Rollout, Upgrade and Rollback Planning

###### How can a rollout or rollback fail? Can it impact already running workloads?

> A rollout failure looks like kube-proxy on some nodes failing to bind a loopback NodePort (e.g. another host process already owns the port). In that case the listener for that one service does not come up, but every other service and every other node continue to work.

Reviewer 2: Part of the question here is, "is it possible that someone enables this feature and restarts kube-proxy, and kube-proxy fails to start?" (I think the answer is "no"; at worst it might log some errors about not being able to bind sockets, right?)

Author: Yeah users would just get an error, clarified that line a bit.

Rollback has a limited failure mode since it only turns off the userspace proxy.

###### What specific metrics should inform a rollback?

> `kubeproxy_nftables_localhost_nodeport_listeners{ip_family}` should be > 0 on nodes where the operator expects loopback NodePorts to be reachable. If it stays at 0 after the gate is enabled and matching services exist then a rollback may be required.

Reviewer 2: That means it's not working but it doesn't mean something bad is happening that would require a rollback. I think the answer here is "None"

###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?

Yes, manual testing was performed with the following steps:
1. Create an nftables cluster without the `KubeProxyNFTablesLocalhostNodePorts` gate enabled.
1. Create a service with a backend pod.
1. Enable the gate and restart kube-proxy; verify calls to localhost NodePort succeed.
1. Disable the gate and restart kube-proxy; verify calls to localhost NodePort fail.
1. Re-enable the gate and restart kube-proxy; verify calls to localhost NodePort succeed.

###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?

No.

### Monitoring Requirements

###### How can an operator determine if the feature is in use by workloads?

> Not possible currently. `kubeproxy_nftables_localhost_nodeport_listeners{ip_family}` shows that kube-proxy has bound loopback listeners, but it does not by itself prove that any workload is connecting to them.

Reviewer 2: But the feature can only be in use at all if the operator has enabled it by changing the kube-proxy configuration.

Author: Ah gotcha, updated, I think I understand the intent of the question better now

###### How can someone using this feature know that it is working for their instance?

- [X] Other (treat as last resort)

>   - Details: An operator can scrape `kubeproxy_nftables_localhost_nodeport_listeners{ip_family}` from kube-proxy on a node and confirm it equals the number of TCP NodePort services they expect to be reachable on loopback.

Reviewer 2: This question is about end users, not operators.

The user can know that it is working by trying to connect to their service on localhost.

###### What are the reasonable SLOs (Service Level Objectives) for the enhancement?

> N/A

Reviewer 2: We probably need to talk about this a little, since we expect the userspace proxy to be noticeably slower than the nftables proxy.

I guess we can say that we expect [time-to-first-packet](https://github.com/kubernetes/community/blob/main/sig-scalability/slos/first_packet_latency.md) and [throughput](https://github.com/kubernetes/community/blob/main/sig-scalability/slos/throughput.md) to be similar to the old userspace kube-proxy, which we don't have any specific measurements of, but it wasn't too awful and people survived. It may be worth saying explicitly that we don't expect people to have very-high-throughput requirements for localhost nodeports.

###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?

- [X] Metrics
  - Metric name: `kubeproxy_nftables_localhost_nodeport_listeners`

>   - Components exposing the metric: `kube-proxy`

Reviewer 2: That doesn't really indicate its health. I think the answer here is that there aren't any.

-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

This KEP proposes implementing a userspace proxy to enable localhost NodePort services across all proxy backends.
The iptables proxy already works with localhost NodePorts; however, it requires enabling a dangerous permission,
`route_localnet`, on each node to allow the kernel to rewrite traffic to localhost. `route_localnet` is
incompatible with IPv6 as well. By rewriting traffic directly at the userspace level instead, no extra
permissions are set, and localhost NodePorts will work across all Linux proxy backends using either IP family.

## Motivation

`route_localnet` is a dangerous kernel permission that kube-proxy in iptables mode sets on every node using IPv4
by default. Implementing a userspace proxy will enable a secure default without breaking existing use cases.
See [#132955](https://github.com/kubernetes/kubernetes/issues/132955).

Kubernetes would like to change the default proxy backend to nftables. Enabling localhost NodePort services

> with nftables will make it easier for the community to accept nftables as the new default.

Reviewer 3: just a side comment that I didn't find where to plug, is that docker also went through this problem and https://github.com/docker/docs/issues/17312#issuecomment-1547368847 it defaults the portmap proxy to userspace https://docs.docker.com/engine/network/port-publishing/ IIRC because of this

### Goals

- Enable localhost NodePort across all Linux kube-proxy modes, for both IPv4 and IPv6, without `route_localnet`
- Stop enabling `route_localnet`

### Non-Goals

- Implement UDP or other non-TCP protocols in the userspace proxy

## Proposal

Add a userspace TCP-only proxy to "kube-proxy" that listens on the loopback address (`127.0.0.1` / `::1`) for
each NodePort and forwards connections to a service's endpoints. Each backend proxy (iptables, nftables, ipvs) instantiates this proxy when `--nodeport-addresses` does
not match all IPs and contains a loopback CIDR, and reconciles its set of listeners on every sync.

> `--nodeport-addresses` will allow `primary` to be part of a list with other IPs and keywords instead of enforcing that it must be a singular value.

Reviewer 3: puff, my memory is not as before, I had to reread https://github.com/kubernetes/kubernetes/pull/123105 about that `primary` value

A new `localhost` keyword will be introduced for `--nodeport-addresses` as a convenience alias for the loopback
addresses of the node's IP families. Operators enable localhost NodePorts by including it alongside other
specific addresses (e.g. `--nodeport-addresses=primary,localhost`).

The userspace proxy is only activated when `--nodeport-addresses` does not match all IPs. When the value
matches all IPs (the zero CIDR, which is the default for iptables and ipvs), the existing behavior
is preserved and the new code path is never entered. This means no default behavior changes for any backend:
operators must explicitly set a non-all-IPs value that includes loopback or `localhost` to opt in. 

> Setting `0.0.0.0/0,localhost` will not activate the proxy as it still includes all IPs. 

Reviewer 3: this looks very ambiguous, can we just fail on validation and do not allow this combination?

Reviewer 2: We _could_, though we allow things like `0.0.0.0/0,127.0.0.1/32` now...

If you don't like those semantics, a better alternative might be to say that `0.0.0.0/0,localhost` _does_ trigger the userspace proxy. That is, the rule is not "if you listen on all IPs, you don't get the userspace proxy", it's "if you don't explicitly request localhost IPs, you don't get the userspace proxy". (This would require more rewriting of the proxy implementations, because we'd have to add code to do userspace proxying to both the "listen on all IPs" path and the "listen on specific IPs" path.)

(EDIT: ah, you suggest the same thing below)

Reviewer 2: Also, if we're adding `localhost`, we should probably add `all` too...

nftables defaults to `primary` and its default will not change; operators who want localhost NodePorts on
nftables must explicitly include `localhost` when setting `--nodeport-addresses`.

iptables IPv4 will continue to use `route_localnet` until this feature graduates to GA, then transition to

> the userspace proxy. The field `--iptables-localhost-nodeports` will be deprecated in favor of explicitly

Reviewer 3: history of this flag here https://github.com/kubernetes/kubernetes/pull/108250

setting `--nodeport-addresses` to include or exclude localhost. Once this feature is stable, iptables IPv4 
will activate the userspace proxy when `--nodeport-addresses` includes localhost or matches all IPs. This differs from other backends, 

> where the userspace proxy is never activated when all IPs are matched.

Reviewer 3: So I think we will have something like

| Proxy Backend | Default `--nodeport-addresses` | Activates Userspace Proxy by Default? | How to Enable / Keep Enabled | How to Disable |
| :--- | :--- | :--- | :--- | :--- |
| **iptables IPv4** | `0.0.0.0/0` (Matches all) | **Yes** | Leave as default, or include `localhost` | Set explicit IP list *without* `localhost` |
| **iptables IPv6** | `::/0` (Matches all) | **No** | Set to e.g., `primary,localhost` | Leave as default, or exclude `localhost` |
| **IPVS (v4/v6)** | `0.0.0.0/0` / `::/0` | **No** | Set to e.g., `primary,localhost` | Leave as default, or exclude `localhost` |
| **nftables (v4/v6)**| `primary` | **No** | Set to `primary,localhost` | Leave as default, or exclude `localhost` |

 
If a user has a node with 3 different network interfaces and wants NodePorts exposed on all of them plus the localhost userspace proxy, I think that this impossible without hardcoding all 3 IPs, and this is not always possible.

If we allow `--nodeport-addresses=0.0.0.0/0,localhost` to mean:

"Let the kernel datapath (iptables/ipvs/nftables) bind to all external IPs (0.0.0.0/0), AND stsart the userspace proxy specifically for loopback."

we give users exactly what they want without forcing them to drop the 0.0.0.0/0 catch-all for their external interfaces.

Reviewer 2: > without hardcoding all 3 IPs

without hardcoding all 3 _CIDR ranges_. That's the whole reason we made `--nodeport-addresses` take CIDRs rather than IPs; because it works much better when deploying a single config to all nodes.

(Aside: I had been thinking about adding labels to `node.status.addresses` in the wake of https://github.com/kubernetes/kubernetes/pull/137081, and one of the use cases was being able to say `kube-proxy --nodeport-addresses node.k8s.io/ifname=eth2,node.k8s.io/ifname=eth3` or `--nodeport-addresses node.k8s.io/network=backend` or whatever)

### User Stories

#### Story 1

A cluster operator migrating from iptables to nftables discovers their workload depends on localhost
NodePorts. Today the migration breaks that workload; with the userspace localhost NodePort proxy it keeps
working.

#### Story 2

In-cluster registries in disconnected environments are a popular production use case for localhost NodePorts.
The tool [Zarf](https://github.com/zarf-dev/zarf) has a built-in process to start up a registry served over a
localhost NodePort service on an airgapped cluster. The process is distro and container runtime interface (CRI)
agnostic, but only works with IPv4 iptables. Localhost is used to connect to the NodePort service because
popular CRIs such as containerd and cri-o allow insecure localhost connections to registries. Without this,
users would need to edit their host node configuration to either include TLS certificates or allow insecure
connections to node IPs. Enabling localhost NodePorts makes it easier for these users to move from iptables
to nftables.

### Risks and Mitigations

> Once this feature is stable, `route_localnet` will not be set for iptables IPv4 backends. This could break users

Reviewer 3: I'm fine with making this an exception, and this can be a follow up if needed, but then I think we should split the iptables replacement of route_localnet from this KEP, there are things like syslog or metrics that listen in udp, users may optin to the new behavior but by default they should be able to keep using route_localnet ... or at least UDP, is not as complex to implement

that is unlikely to affect a significant portion of users. If this is deemed high enough impact,
the proxy could be extended to support the other protocols.

## Design Details

### Service Feature Support

The userspace localhost NodePort proxy is a Layer 4 TCP forwarder. The following covers Service spec fields and how the proxy handles them.

**Implemented:**

- `protocol: TCP` — the proxy creates a listener for every TCP NodePort included in the sync.
- `sessionAffinity: ClientIP` — the proxy pins traffic to a single endpoint for `sessionAffinityConfig.clientIP.timeoutSeconds`. Because all connections originate from `127.0.0.1`/`::1`, ClientIP affinity effectively pins *all* traffic through the listener to one endpoint—not per external client—until the timeout elapses.
- `externalTrafficPolicy: Local` — the proxy restricts its endpoint pool to node-local endpoints. If no local endpoints exist the listener is removed and connections are refused.
- `trafficDistribution` — implemented, endpoints are already categorized before the userspace proxy is created.

**Not implemented:**

> - `protocol: UDP` / `protocol: SCTP` — the proxy creates no listener and logs an error. For iptables IPv4 this is a regression at GA, once `route_localnet` is removed the userspace proxy does not replace that coverage. These are not implemented as we believe this is a niche use case, however, if users request this feature this can be reconsidered. 

Reviewer 3: the iptables backend seems too much trouble, is there a way we can simplify this KEP to avoid breaking that behavior?
It is really nftables only the one we want to fix, we can also use this for encouraging people to move to teh new backend

Reviewer 2: My argument for implementing this in iptables too is that it will be a while before we are ready to deprecate iptables mode, but I would like to deprecate setting `route_localnet` sooner than that if we can, to improve security for the people who haven't migrated away from iptables yet.

If not for that, I'd agree that we should just only worry about nftables, to simplify things.

Reviewer 3: I think simplifying is better , I share the temptation of killing two birds with one stone, but I feel that the iptables mode migration , if causes regressions, it will overcomplicate later the graduation or backports
